comms/src/connection/connection.rs

```diff
@@ -340,9 +355,32 @@ impl EstablishedConnection {
     }
 }
 
-impl From<zmq::Socket> for EstablishedConnection {
-    fn from(socket: zmq::Socket) -> Self {
-        Self { socket }
+/// Extract the actual address that we're connected to. On inbound connections, once can delegate port selection to
+/// the OS, (e.g. "127.0.0.1:0") which means that the actual port we're connecting to isn't known until the binding
+/// has been made. This function queries the socket for the connection info, and extracts the address & port if it
+/// was a TCP connection, returning None otherwise
+fn get_socket_address(socket: &zmq::Socket) -> Option<SocketAddress> {
+    let addr = match socket.get_last_endpoint() {
+        Ok(v) => v.unwrap(),
+        Err(_) => return None,
+    };
+    let parts = &addr.split("//").collect::<Vec<&str>>();
+    if parts.len() < 2 || parts[0] != "tcp:" {
+        return None;
+    }
+    let addr = parts[1];
+    SocketAddress::from_str(&addr).ok()
```

Comment:
For another PR, but it would be useful to have a NetAddress returned for onion and i2p addresses

Reply:
Agreed. I'm sure there are edge cases to catch in this function